File: PythonVersion/example/MyTest.py
from DataStructure.PythonVersion.source.MyList import MyList
from DataStructure.PythonVersion.source.MyMap import MyMap
from DataStructure.PythonVersion.source.MyQueue import MyQueue
from DataStructure.PythonVersion.source.MyStack import MyStack
from DataStructure.PythonVersion.source.MyTree import MyTree


def test_list():
    mylist = MyList[int]()
    mylist.add(111)
    mylist.add(222)
    mylist.add(333)
    mylist.add(444)
    mylist.add(555)
    mylist.add(666)
    mylist.add(777)
    assert len(mylist) == 7
    assert mylist.get(1) == 222
    mylist.remove(1)
    assert len(mylist) == 6
    assert mylist.get(1) == 333
    mylist.set(1, 234)
    assert mylist.get(1) == 234
    mylist.insert(1, 890)
    assert mylist.get(1) == 890
    assert len(mylist) == 7
    assert str(mylist) == '[111, 890, 234, 444, 555, 666, 777]'
    mylist.clear()
    assert mylist.is_empty()

def test_map():
    mymap = MyMap[str, str]()
    mymap.put("key1", "value1")
    mymap.put("key2", "value2")
    mymap.put("key3", "value3")
    mymap.put("key4", "value4")
    mymap.put("key5", "value5")
    assert len(mymap) == 5
    assert mymap.get("key4") == "value4"
    assert mymap.get_or_default("key7", "value7") == "value7"
    mymap.replace("key3", "value6")
    assert str(mymap) == '{key1=value1, key2=value2, key3=value6, key4=value4, key5=value5}'
    assert mymap.contains_key("key5")
    mymap.remove("key1")
    assert str(mymap) == '{key2=value2, key3=value6, key4=value4, key5=value5}'
    mymap.clear()
    assert mymap.is_empty()

def test_stack():
    mystack = MyStack[int]()
    mystack.push(1)
    mystack.push(2)
    mystack.push(3)
    assert len(mystack) == 3
    mystack.push(4)
    mystack.push(5)
    assert len(mystack) == 5
    assert str(mystack) == '[1, 2, 3, 4, 5]'
    assert mystack.pop() == 5
    assert mystack.pop() == 4
    assert mystack.pop() == 3
    assert len(mystack) == 2
    assert mystack.peek() == 2
    assert len(mystack) == 2

def test_queue():
    myqueue = MyQueue[int]()
    myqueue.enqueue(1)
    myqueue.enqueue(2)
    myqueue.enqueue(3)
    assert len(myqueue) == 3
    myqueue.enqueue(4)
    myqueue.enqueue(5)
    assert len(myqueue) == 5
    assert str(myqueue) == '[1, 2, 3, 4, 5]'
    assert myqueue.dequeue() == 1
    assert myqueue.dequeue() == 2
    assert myqueue.dequeue() == 3
    assert len(myqueue) == 2
    assert myqueue.peek() == 4
    assert len(myqueue) == 2

def test_tree():
    mytree = MyTree[str]()
    mytree.add_leaf("aaa")
    mytree.add_leaf("bbb")
    mytree.add_leaf("ccc")
    mytree.add_leaf("ddd")
    mytree.add_leaf("eee")
    mytree.add_branch()
    mytree.insert_branch(1)
    mytree.insert_leaf(2, "miao")
    assert len(mytree) == 8
    assert str(mytree) == '[aaa, [], miao, bbb, ccc, ddd, eee, []]'
    for i in range(len(mytree)):
        if mytree.is_branch(i):
            mytree.get_branch(i).add_leaf("111")
            mytree.get_branch(i).add_leaf("222")
    assert str(mytree) == '[aaa, [111, 222], miao, bbb, ccc, ddd, eee, [111, 222]]'
    mytree.remove_node(1)
    assert str(mytree) == '[aaa, miao, bbb, ccc, ddd, eee, [111, 222]]'
    assert mytree.is_branch(6)
    mytree2 = mytree.get_branch(6)
    assert len(mytree2) == 2
    assert str(mytree2) == '[111, 222]'
    assert str(mytree2.get_parent()) == "[aaa, miao, bbb, ccc, ddd, eee, [111, 222]]"
    assert mytree.is_leaf(3)
    assert mytree.get_leaf(3) == "ccc"

if __name__ == '__main__':
    test_list()
    test_map()
    test_stack()
    test_queue()
    test_tree()

File: JavaVersion/example/MyTest.java
package javaVersion.example;

import javaVersion.source.*;

public class MyTest {
    public static void main(String[] args) {
        testList();
        testStack();
        testQueue();
        testMap();
        testTree();
    }
    public static void testList() {
        MyList<Integer> array = new MyList<>();
        array.add(111);
        array.add(222);
        array.add(333);
        array.add(444);
        array.add(555);
        array.add(666);
        array.add(777);
        assert array.size() == 7;
        assert array.get(1).equals(222);
        array.remove(1);
        assert array.size() == 6;
        assert array.get(1).equals(333);
        array.set(1, 234);
        assert array.get(1).equals(234);
        array.insert(1, 890);
        assert array.get(1).equals(890);
        assert array.size() == 7;
        assert array.toString().equals("[111, 890, 234, 444, 555, 666, 777]");
        array.clear();
        assert array.toString().equals("[]");
    }
    public static void testQueue() {
        MyQueue<Integer> queue = new MyQueue<>();
        queue.enQueue(1);
        queue.enQueue(2);
        assert queue.size() == 2;
        queue.enQueue(3);
        assert queue.size() == 3;
        assert queue.toString().equals("[1, 2, 3]");
        assert queue.deQueue().equals(1);
        assert queue.deQueue().equals(2);
        assert queue.size() == 1;
        assert queue.peek().equals(3);
    }
    public static void testStack() {
        MyStack<Integer> stack = new MyStack<>();
        stack.push(1);
        stack.push(2);
        assert stack.size() == 2;
        stack.push(3);
        assert stack.size() == 3;
        assert stack.toString().equals("[1, 2, 3]");
        assert stack.pop().equals(3);
        assert stack.pop().equals(2);
        assert stack.size() == 1;
        assert stack.peek().equals(1);
    }
    public static void testMap() {
        MyMap<String, String> map = new MyMap<>();
        map.put("key1", "value1");
        map.put("key2", "value2");
        map.put("key3", "value3");
        assert map.size() == 3;
        assert map.get("key1").equals("value1");
        assert map.getOrDefault("key4", "value4").equals("value4");
        map.replace("key2", "value5");
        assert map.get("key2").equals("value5");
        assert map.toString().equals("{key1=value1, key2=value5, key3=value3}");
        assert map.containKey("key3");
        map.remove("key1");
        assert map.toString().equals("{key2=value5, key3=value3}");
        map.insert(1, "key6", "value6");
        assert map.toString().equals("{key2=value5, key6=value6, key3=value3}");
        map.clear();
        assert map.toString().equals("{}");
    }
    public static void testTree() {
        MyTree<String> tree = new MyTree<>();
        tree.addLeaf("aaa");
        tree.addLeaf("bbb");
        tree.addLeaf("ccc");
        tree.addLeaf("ddd");
        tree.addLeaf("eee");
        tree.addBranch();
        System.out.println(tree);
        tree.insertBranch(1);
        System.out.println(tree);
        tree.insertLeaf(2, "miao");
        System.out.println(tree);
        assert tree.size() == 8;
        assert tree.toString().equals("[aaa, [], miao, bbb, ccc, ddd, eee, []]");
        for(int i = 0; i < tree.size(); i++) {
            if(tree.isBranch(i)){
                tree.getBranch(i).addLeaf("111");
                tree.getBranch(i).addLeaf("222");
            }
        }
        assert tree.size() == 8;
        assert tree.toString().equals("[aaa, [111, 222], miao, bbb, ccc, ddd, eee, [111, 222]]");
        tree.removeNode(1);
        assert tree.toString().equals("[aaa, miao, bbb, ccc, ddd, eee, [111, 222]]");
        assert tree.isBranch(6);
        MyTree<String> tree2 = tree.getBranch(6);
        assert tree2.size() == 2;
        assert tree2.toString().equals("[111, 222]");
        assert tree2.getParent().toString().equals("[aaa, miao, bbb, ccc, ddd, eee, [111, 222]]");
        assert tree.getLeaf(3).equals("ccc");
        assert tree.isLeaf(2);
    }
}
